updated node version, added node list style rules

aaron-prindle opened the pull request.

dlorenc commented on reproducible/ubuntu/nodebootstrap.bzl:
```diff
@@ -56,7 +56,7 @@ docker rm $cid
     attrs = {
         "_builder_image": attr.label(
-            default = Label("//reproducible/ubuntu:nodejs_builder.tar"),
+            default = Label("//reproducible/ubuntu:node_8_5_0_builder.tar"),
```
Why is this one hardcoded? Should we leave off a default to prevent people from making mistakes?

dlorenc commented on reproducible/ubuntu/BUILD:
```diff
@@ -55,16 +55,22 @@ docker_build(
 )
 
-# nodejs_builder is a docker image that is capable of building nodejs running
+# node_%s_builder is a docker image that is capable of building nodejs running
```
nit: move this comment under the map and back on top of the rule.

aaron-prindle commented:
requested changes made.  properly parameterized the job for bazel, renamed nodebootstrap img to node_gen_tar for clarity.  but the builder image creation into the node_gen_tar script as it is not something that should be built directly.

dlorenc commented on reproducible/ubuntu/node_gen_tar.bzl:
```diff
@@ -73,17 +73,23 @@ load(
 )
 
-def nodebootstrap_image(name, env=None):
+def node_gen_tar_rule(name, env=None):
```
nit: make this node_gen_tar and make the other one _node_gen_tar

aaron-prindle replied:
done

dlorenc commented:
Travis looks like a flake, once you do the minor rename it should pass.

aaron-prindle commented:
I believe the issue with Travis is that it is building a lot images due to the 'make test' lines:
	bazel build //...
  	bazel test --test_output=errors //...
I believe that the tests are timing out due to length

aaron-prindle commented:
added a check to find the number of cores on host machine and use that many in nodejs compilation.  Also removed the .tar generation as seen here: https://github.com/GoogleCloudPlatform/debian-docker/pull/115

aaron-prindle commented:
split the rules out.  updated to use checked-in, compiled node.tar.gz to avoid having to compile in Travis on every PR

dlorenc commented on WORKSPACE:
```diff
@@ -98,8 +98,7 @@ http_file(
-)
+NODE_MAP = {
+    "8.5.0": "dd5ac97074882ba62892bdb91ec5e624c338d772a6218da08c182231031a4446",
+    # "8.6.0": "80551d47440c602c5825a3f6611e13cb0c57d277234dbff693af433a88351a08",
```
uncommenrt?

dlorenc replied:
actually, delete this variable?

aaron-prindle replied:
Can you provide args to a bazel build?  My understanding is that I would need a list and a loop to generate all available node build options.

aaron-prindle replied:
This should be deleted from WORKSPACE, is there an alternative to the map existing for reproducible/ubuntu though?

dlorenc commented on dockerfile_build/Dockerfile.ubuntu_build:
```diff
@@ -2,6 +2,8 @@ FROM bazel/reproducible:ubuntu
     apt-get install --no-install-recommends -y -q \
       build-essential \
+      curl \
+      ca-certificates \
```
We should get this in the base image. Can you open a bug to track that for now?

aaron-prindle replied:
What base image are you referring to?  This is the 'ubunut+build_essential, etc.' base image.  Are you referring to the 'ubuntu' image?

dlorenc replied:
ubuntu

aaron-prindle replied:
Issue here:
https://github.com/GoogleCloudPlatform/debian-docker/issues/117

dlorenc commented on reproducible/ubuntu/BUILD:
```diff
@@ -24,47 +24,39 @@ docker_build(
-load("//reproducible/ubuntu:nodebootstrap.bzl", "nodebootstrap_image")
+NODE_MAP = {
+    "8_5_0": "dd5ac97074882ba62892bdb91ec5e624c338d772a6218da08c182231031a4446",
+    # "8_6_0": "80551d47440c602c5825a3f6611e13cb0c57d277234dbff693af433a88351a08",
```
uncomment 8.6?

aaron-prindle replied:
This was commented as a "stub".  Should we support these two versions of node?  To build both, we would need both tarballs checked in.

dlorenc commented on reproducible/ubuntu/mknodeimage.sh:
```diff
@@ -1,19 +1,29 @@
 
+VERSION=$1
+
 ln -s /usr/bin/python2.7 /usr/bin/python
```
Some comments in this file would help to explain how it works, what arguments it takes and what it does.

dlorenc commented on reproducible/ubuntu/mknodeimage.sh:
```diff
@@ -1,19 +1,29 @@
 # pass -n to gzip to strip timestamps
 # strip the '.' with --transform thatp tar includes at the root to build nodejs
-GZIP="-n" tar --numeric-owner -czf /workspace/nodejs.tar.gz -C "$OUTPUTDIR" . --transform='s,^./,,' --mtime='1970-01-01'
+GZIP="-n" tar --numeric-owner -czf /workspace/node_${VERSION}_compiled.tar.gz -C "$OUTPUTDIR" . --transform='s,^./,,' --mtime='1970-01-01'
```
not sure you need the transform= part, I think that's only important for rootfs tarballs.

aaron-prindle replied:
The node installation is a rootfs tarball created from 'make install' into a separated dir.

dlorenc commented on reproducible/ubuntu/node_gen_tar.sh:
```diff
@@ -0,0 +1,31 @@
+NODE_OUTPUT_PATH=$2
+
+# TODO(aaron-prindle) change this to use the exec method
+docker load -i $(pwd)/dockerfile_build/ubuntu_build.tar | awk '{print $3}'
```
I don't think you'll be able to change this to exec - ubuntu_build.tar is an actual tar with no incremental loader.

aaron-prindle replied:
todo removed

dlorenc commented on reproducible/ubuntu/node_gen_tar.sh:
```diff
@@ -0,0 +1,31 @@
+
+# TODO(aaron-prindle) change this to use the exec method
+docker load -i $(pwd)/dockerfile_build/ubuntu_build.tar | awk '{print $3}'
+IMG_NAME=$(docker load -i $(pwd)/dockerfile_build/ubuntu_build.tar | awk '{print $3}')
```
Do you need to load this twice?

aaron-prindle replied:
no, removed

dlorenc commented on reproducible/ubuntu/node_gen_tar.sh:
```diff
@@ -0,0 +1,31 @@
+IMG_NAME=$(docker load -i $(pwd)/dockerfile_build/ubuntu_build.tar | awk '{print $3}')
+
+# Run the base image.
+CID=$(docker run -d --privileged -v $(pwd)/reproducible/ubuntu/mknodeimage.sh:/mknodeimage.sh $IMG_NAME /mknodeimage.sh $VERSION)
```
I don't think you need privileged here.

aaron-prindle replied:
done
